sql/expression/function/aggregation/window/first_value.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2021 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package window
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/dolthub/go-mysql-server/sql/expression"
+
+	"github.com/dolthub/go-mysql-server/sql"
+)
+
+type FirstValue struct {
+	window *sql.Window
+	expression.UnaryExpression
+	pos int
+}
+
+var _ sql.FunctionExpression = (*FirstValue)(nil)
+var _ sql.WindowAggregation = (*FirstValue)(nil)
+
+func NewFirstValue(ctx *sql.Context, e sql.Expression) sql.Expression {
+	return &FirstValue{nil, expression.UnaryExpression{Child: e}, 0}
+}
+
+// Window implements sql.WindowExpression
+func (f *FirstValue) Window() *sql.Window {
+	return f.window
+}
+
+// IsNullable implements sql.Expression
+func (f *FirstValue) Resolved() bool {
+	return windowResolved(f.window)
+}
+
+func (f *FirstValue) NewBuffer() sql.Row {
+	return sql.NewRow(make([]sql.Row, 0))
+}
+
+func (f *FirstValue) String() string {
+	sb := strings.Builder{}
+	sb.WriteString(fmt.Sprintf("first_value(%s)", f.Child.String()))
+	if f.window != nil {
+		sb.WriteString(" ")
+		sb.WriteString(f.window.String())
+	}
+	return sb.String()
+}
+
+func (f *FirstValue) DebugString() string {
+	sb := strings.Builder{}
+	//sb.WriteString(fmt.Sprintf("first_value(%s)", f.Child.String()))
+	if f.window != nil {
+		sb.WriteString(" ")
+		sb.WriteString(sql.DebugString(f.window))
+	}
+	return sb.String()
+}
+
+// FunctionName implements sql.FunctionExpression
+func (f *FirstValue) FunctionName() string {
+	return "FIRST_VALUE"
+}
+
+// Type implements sql.Expression
+func (f *FirstValue) Type() sql.Type {
+	return f.Child.Type()
+}
+
+// IsNullable implements sql.Expression
+func (f *FirstValue) IsNullable() bool {
+	return false
+}
+
+// Eval implements sql.Expression
+func (f *FirstValue) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	panic("eval called on window function")
+}
+
+// Children implements sql.Expression
+func (f *FirstValue) Children() []sql.Expression {
+	if f == nil {
+		return nil
+	}
+	return append(f.window.ToExpressions(), f.Child)
+}
+
+// WithChildren implements sql.Expression
+func (f *FirstValue) WithChildren(ctx *sql.Context, children ...sql.Expression) (sql.Expression, error) {
+	nf := *f
+	window, err := f.window.FromExpressions(children[:len(children)-1])
+	if err != nil {
+		return nil, err
+	}
+
+	nf.Child = children[len(children)-1]
+	nf.window = window
```

Comment:
You should just use WithWindow()

Reply:
i get the point of the helper function if I didn't have a child expr, but isn't it a bit awkward to convert back to a pointer and handle an error just to call a function that sets window?